linkerd/protocol/base-http/src/main/scala/io/buoyant/linkerd/http/ZipkinTracePropagator.scala

```diff
@@ -0,0 +1,248 @@
+
+    caseInsensitiveGet(headers, ZipkinB3SingleHeader) match {
+      case Some(v) => {
+        val (traceId, sampled, flags) = getFromB3SingleHeader(v)
+        traceId
+      }
+      case _ => getFromXB3MultiHeaders(headers)
+    }
+  }
+
+  /**
+   * There's no way to know what the downstream is capable of, and it is more likely it supports "X-B3-*" vs not, so,
+   * the portable choice is to always write down "X-B3-"  even if we read "b3". Finagle does a similar thing in
+   * finagle-http-base: read "b3", write back "X-B3-".
+   *
+   * @note This method does not touches the request, b3 header is not remove, x-b3- headers not removed
+   * @note The "b3" encodings specs, single or multi: https://github.com/openzipkin/b3-propagation#http-encodings
+   * @param headers
+   * @return Option[TraceId]
+   */
+
+  def set[H: HeadersLike](headers: H, id: TraceId): Unit = {
+    val headersLike = implicitly[HeadersLike[H]]
+
+    // remove any b3 single header, if present, before setting x-b3 multi headers
+    val _ = headersLike.remove(headers, ZipkinB3SingleHeader)
+
+    headersLike.set(headers, ZipkinSpanHeader, id.spanId.toString)
+
+    // support setting a 128bit traceid
+    if (id.traceIdHigh.isEmpty) {
+      headersLike.set(headers, ZipkinTraceHeader, id.traceId.toString)
+    } else {
+      headersLike.set(headers, ZipkinTraceHeader, id.traceIdHigh.get.toString + id.traceId.toString)
+    }
+
+    headersLike.set(headers, ZipkinParentHeader, id.parentId.toString)
+
+    //the only valid value for x-b3-flags is 1, which means debug
+    if (id.flags.isDebug) {
+      //when setting x-b3-flags to debug, 1, do not also set x-b3-sampled because it is a redundant info
+      headersLike.set(headers, ZipkinFlagsHeader, id.flags.toLong.toString)
+    } else {
+      id.sampled match {
+        // valid values fro x-b3-sampled are 1 and 0
+        case Some(true) => headersLike.set(headers, ZipkinSampleHeader, "1")
+        case Some(false) => headersLike.set(headers, ZipkinSampleHeader, "0")
+        case None => // do nothing
+      }
+    }
+
+    ()
+  }
+
+  def getSampler[H: HeadersLike](headers: H): Option[Float] = {
+    val headersLike = implicitly[HeadersLike[H]]
+    val samplerNone: Option[Float] = None
+    val samplerTrue: Option[Float] = Option(1.0f)
+    val samplerFalse: Option[Float] = Option(0.0f)
+
+    // first try getting flags/sampled info from b3 single header
+    caseInsensitiveGet(headers, ZipkinB3SingleHeader) match {
+      case Some(v) => {
+        val (traceId, sampled, flags) = getFromB3SingleHeader(v)
+        flags match {
+          case Flags(1) => samplerTrue // flags debug
+          case _ => { // flags invalid or not present, look for sampled field
+            sampled match {
+              case Some(true) => samplerTrue
+              case Some(false) => samplerFalse
+              case _ => samplerNone
+            }
+          }
+        }
+      }
```

Comment:
If you want to flatten/simplify your match as we spoke, I think you can have some kind an ADT that encodes all the variants that can be returned from `getFromB3SingleHeader()` . Then you can have something like 

getFromB3SingleHeader(v) match {
   case Variant1(x,y) => ....
   case Variant2(z) => ......
   etc, etc.... 
}